### ISSUE TITLE
(BUG) Shade tries to move before you’ve selected a position 

### ISSUE THREAD

**Issue by JJJ** [tonesto7/homebridge-smartthings-v2#82]
-  [x] That _OAuth_ is Enabled for the SmartApp under the IDE.
-  [x] The SmartApp and Device Handler are using the latest code available.
-  [x] That Both the SmartApps and Device Handlers have been _Published for You_ in the IDE.

---
## About Your Setup
-  How many devices are detected?: 3
-  Mobile App Version(Not required): 1.6.42-350
-  SmartApp Version: 2.2.1
-  Device Handler Version: 2.2.1
-  Homebridge Version: 0.4.46
-  NodeJS Version: 12

## Expected Behavior
When moving this slider, I expect for the shade to not start moving until I’ve lifted my finger from the slider:

![CAAB2E74-73B3-4A82-93FA-BF5C16FDDA70](https://user-images.githubusercontent.com/88951/72773097-afad6e00-3bcb-11ea-8b35-20c61a2b0585.png)

## Current Behavior
When I invoke this slider UI, no matter what position I attempt to pick, the shade motor will raise or lower the shade about 4 inches, but very slowly and painfully sounding.

## Steps to Reproduce (for bugs)
1. Setup a shade
2. Confirm that open and close work OK
3. Pull up the slider and try to set a specific position
4. See that your shade prematurely tries to move before you’ve finished setting a position

## Context
This issue existed in v1 of this plugin, so this is not a regression, though I’d hoped it would have magically fixed itself for v2.

Normally I can just use automations or buttons to raise and lower the shades, so this bug does not make the shades completely unusable. It just makes it so you need to avoid using that slider control.

---

I will happily dig through my logs and look for any relevant entries. 🙏

**Comment by tonesto7** [comment]
yea I probably need to add debouncing to the commands.  The way homekit works it send the values for every move of the slider.

**Comment by tonesto7** [comment]
I’m sorry I haven’t got to this yet.  
My wife’s birthday is coming up and I’m traveling for work all next week so I’m prepping.  
I will try to get this in before I leave tomorrow

**Comment by JJJ** [comment]
@tonesto7 not urgent. Happy birthday to your wife! 🎂

**Comment by tonesto7** [comment]
@Areson Can you take a look at the logic under ST_Client.js for the debounce... 
I can't get it to actually debounce on level type values.

**Comment by tonesto7** [comment]
@JJJ I am trying to work on this.  
I'm trying to get this working with all slider based commands but the debounce is being ignored and fires repeatedly anyways

**Comment by JJJ** [comment]
@tonesto7 Ain’t nothing easy in this world. Thanks for keeping this issue updated. If I can find time, I may also dig in and take a look at it.

**Comment by Areson** [comment]
@tonesto7 I'll try and take a look at this today.

**Comment by tonesto7** [comment]
Thanks... I've been playing with it so make sure you pull from the latest commits

**Comment by tonesto7** [comment]
@Areson I think it's working better now. I increased the wait to 1500ms and it seems to be better.
Maybe you can make it work smoother.

I hate adding these delays but it should reduce the bursts of successive calls.
Maybe I will add in delay settings to the config to allow users to fine-tune them.

**Comment by Areson** [comment]
@tonesto7 It's a tough nut to crack since we don't get "finger up" notifications. I was playing around with this when I added support for my Sonos and the longer the wait the less issues you have, but it makes the experience that much less smooth. Ultimately it's a trade off between how smooth you want the hardware response to be vs. how long you think folks are willing to wait.

The other thing I would mention is that when I implemented it I ended up putting debouncing into the event notification handling rather than sending the commands over the wire. I had the handler remember the last value it received and only queued up an update to send if the value had changed and after a delay to debounce. Smoothing the chatter coming from the services seemed to work a bit better for me. The downside is that you have to add some code to each of the specific services that you want handled that way.

**Comment by JJJ** [comment]
> The other thing I would mention is that when I implemented it I ended up putting debouncing into the event notification handling rather than sending the commands over the wire. I had the handler remember the last value it received and only queued up an update to send if the value had changed and after a delay to debounce. Smoothing the chatter coming from the services seemed to work a bit better for me. The downside is that you have to add some code to each of the specific services that you want handled that way.

This would probably help a few different places, in my opinion. No reason to go through the entire event loop if the value isn't even being changed from what it's set to currently.

If you wanted to ship the 1500ms delay, I'm happy to give it a go.

**Comment by tonesto7** [comment]
I have already built out a new command system that passes all commands through it and only implements the delays for certain commands.
I think I have it working smoothly.  I will try to get the new version out today

**Comment by tonesto7** [comment]
@JJJ 

Ok, so I published a beta version of the plugin!
```npm i -g homebridge-smartthings-v2@beta```

Make sure you are running the latest SmartApp code as well

**Comment by tonesto7** [comment]
@JJJ Did you have a chance to test out the ```beta``` version of the plugin?

**Comment by JJJ** [comment]
@tonesto7 not yet, I will try today and report back.

**Comment by JJJ** [comment]
@tonesto7 tested!

It is vastly improved, and I think works well enough to ship as is. 💯

It isn't exactly _perfect_, but I'm also a bit of a stickler. 😆

There is still a weird thing that can happen with the slider control. If I move it to a position on or near completely open or closed and let go, sometimes it bounces back to the current position, and then catches back up after a few seconds.

I would call _that_ an enhancement to be improved later when there are less important things to do. Knowing HomeKit, this might not even be something you make much better than this right now.

Thanks for improving this!

**Comment by Areson** [comment]
If the debouncing was moved inside the event handler like I suggested, then
depending on how it was done this has a side effect that Homekit isn't
updated until after the debouncing timeout expires. If your device requests
a value from Homekit prior to that it may be the old value which is why it
jumps. Once the timeout expires and the new value is put into place then
the next read would be the proper value.

**Comment by tonesto7** [comment]
@Areson I did listen to you 😃!  I just didn’t want to add duplicate code to every characteristic so I took my implementation in a slightly different direction which I think works really well and seems to be extremely fast.
I use event listeners now to relay commands now.  Meaning that when homekit sends the command it receives the callback immediately which sets the characteristic value.
I’m definitely open to your input and code changes you submit.

@JJJ I know what issue you are referring to, and it’s a rare one that wasn’t worth me coding in the debounce logic into every command listener. 

I’ve been extremely happy with the responsiveness of using Siri with HomeKit lately as commands are executed much quicker now with the plugin.
